### PR TITLE
Linear regression subpackage added to linear dynamical systems

### PR DESCRIPTION
**Pull request by ncguilbeault**
This PR adds classes and workflows to the Bonsai.ML.LinearDynamicalSystems package to enable the Kalman filter to perform online linear regression. Additionally, several new visualisers were added to the Bonsai.ML.Visualizers package to allow users to visualise multidimensional arrays using OxyPlot's Heatmap.

# Linear Regression Subpackage
3 new Bonsai nodes have been added. The `KFModelParameters` node provides a class which will create the parameters needed to initialise the Kalman filter as a linear regression model. The `GridParameters` and `MultivariatePDF` nodes provide methods to modify and access the python model attributes that relate to the creation of a multivariate distribution.

3 new workflows have also been added. They allow for model creation, Bayesian inference, and computing the multivariate pdf.

# Linear Dynamical Systems Package
2 new nodes were added for manipulating 2D multidimensional arrays of double. The `Slice` operator will subsample an array using the specified start/stop indices of the desired columns/rows. The `Reshape` operator allows multidimensional arrays to be reshaped into a different number of rows/columns.

# Visualizers Package 
The `HeatMapSeriesOxyPlotBase` class provides underlying access to the `HeatMapSeries` from the OxyPlot package. Both the `MultivariatePDFVisualizer` and the `MultidimensionalArrayVisualizer` utilise this base class for visualising data.

### CODE REVIEW

File: src/Bonsai.ML.Visualizers/HeatMapSeriesOxyPlotBase.cs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+            };
+
+            model.Axes.Add(colorAxis);
+            model.Series.Add(heatMapSeries);
+
+            view.Model = model;
+            Controls.Add(view);
+
+            InitilizeColorPalette();
+            InitilizeRenderMethod();
```
Comment (glopesdev) [inline src/Bonsai.ML.Visualizers/HeatMapSeriesOxyPlotBase.cs:101]:
```suggestion
            InitializeColorPalette();
            InitializeRenderMethod();
```

This class is internal so not critical, but seems like the names have typos here.

File: src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+{
+
+    /// <summary>
+    /// Provides a type visualizer to display the state components of a Kalman Filter kinematics model.
+    /// </summary>
+    public class MultidimensionalArrayVisualizer : DialogTypeVisualizer
+    {
+
+        private int paletteSelectedIndex = 0;
+        private int renderMethodSelectedIndex = 0;
```
Comment (glopesdev) [inline src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs:28]: Looks like these could be removed and leave only the corresponding properties, e.g. with `{ get; set; }`.

File: src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        /// <summary>
+        /// Size of the window when loaded
```
Comment (glopesdev) [inline src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs:31]:
```suggestion
        /// Gets or sets the size of the window when loaded.
```

Documentation convention on properties is to indicate which operations are available (get/set/ or both).

Also the `.layout` files already take care of memorizing the layout window size. Is there a reason we need to override that behavior?

File: src/Bonsai.ML.Visualizers/MultivariatePDFVisualizer.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        /// <summary>
+        /// Size of the window when loaded
+        /// </summary>
+        public Size Size { get; set; } = new Size(320, 240);
+
+        /// <summary>
+        /// The selected index of the color palette to use
+        /// </summary>
+        public int PaletteSelectedIndex { get => paletteSelectedIndex; set => paletteSelectedIndex = value; }
+        public int RenderMethodSelectedIndex { get => renderMethodSelectedIndex; set => renderMethodSelectedIndex = value; }
```
Comment (glopesdev) [inline src/Bonsai.ML.Visualizers/MultivariatePDFVisualizer.cs:37]: Same comments as above.

File: src/Bonsai.ML.Visualizers/StateComponentVisualizer.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    {
+
+        private DateTime? _startTime;
+
+        private TimeSeriesOxyPlotBase Plot;
+
+        /// <summary>
+        /// Size of the window when loaded
+        /// </summary>
+        public Size Size { get; set; } = new Size(320, 240);
```
Comment (glopesdev) [inline src/Bonsai.ML.Visualizers/StateComponentVisualizer.cs:28]: Same question as above.

File: src/Bonsai.ML.Visualizers/TimeSeriesOxyPlotBase.cs [hunk 6]
```diff
@@ -41,6 +43,16 @@ internal class TimeSeriesOxyPlotBase : UserControl
...
+        public ToolStripComboBox ComboBox
+        {
+            get => comboBox;
+        }
+
+        public StatusStrip StatusStrip
+        {
+            get => statusStrip;
+        }
+
```
Comment (glopesdev) [inline src/Bonsai.ML.Visualizers/TimeSeriesOxyPlotBase.cs:55]:
```suggestion
        public ToolStripComboBox ComboBox => comboBox;

        public StatusStrip StatusStrip => statusStrip;
```

For getter-only properties it is possible to go even simpler.

**Review by glopesdev — CHANGES_REQUESTED** [review]
Looks really nice overall, just went through everything and found a few inconsistencies on documentation, formatting and possibly naming and design conventions. Otherwise all good!

File: src/Bonsai.ML.LinearDynamicalSystems/Bonsai.ML.LinearDynamicalSystems.csproj [hunk 7]
```diff
@@ -4,7 +4,7 @@
     <Description>A Bonsai package for implementing the Kalman Filter using python.</Description>
     <PackageTags>Bonsai Rx ML KalmanFilter LinearDynamicalSystems</PackageTags>
     <TargetFrameworks>net472;netstandard2.0</TargetFrameworks>
-    <Version>0.1.0</Version>
+    <Version>0.1.1</Version>
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/Bonsai.ML.LinearDynamicalSystems.csproj]: Updated version number

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+namespace Bonsai.ML.LinearDynamicalSystems.LinearRegression
+{
+
+    /// <summary>
+    /// 2D grid parameters used for calculating the PDF of a multivariate distribution
+    /// </summary>
+    [Description("2D grid parameters used for calculating the PDF of a multivariate distribution")]
+    [Combinator()]
+    [WorkflowElementCategory(ElementCategory.Source)]
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs]: I've updated this throughout the classes to use the same template and added better descriptions and XML documentation

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+        private string _y0String;
+        private string _y1String;
+        private string _ystepsString;
+
+        /// <summary>
+        /// Lower bound of the x axis
+        /// </summary>
+        [JsonProperty("x0")]
+        [Description("Lower bound of the x axis")]
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs]: Looks good, I've updated this throughout

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+                _x1String = double.IsNaN(_x1) ? "None" : _x1.ToString();
+            }
+        }
+        
+        /// <summary>
+        /// Number of steps along the x axis
+        /// </summary>
+        [JsonProperty("xsteps")]
+        [Description("Number of steps along the x axis")]
+        public int Xsteps
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs]: Updated

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        {
+            return Observable.Select(source, x =>
+                new GridParameters {
+    				X0 = _x0,
+                    X1 = _x1,
+                    Xsteps = _xsteps,
+    				Y0 = _y0,
+                    Y1 = _y1,
+                    Ysteps = _ysteps,
+                });
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs]: Interesting... Inside of my code editor, everything looks fine

File: src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs (hunk 2 above)
Comment (ncguilbeault) [reply src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs]: Updated this as well

File: src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs (hunk 3 above)
Comment (ncguilbeault) [reply src/Bonsai.ML.Visualizers/MultidimensionalArrayVisualizer.cs]: I've removed the get/set property for Size on the visualizers

File: src/Bonsai.ML.Visualizers/StateComponentVisualizer.cs (hunk 5 above)
Comment (ncguilbeault) [reply src/Bonsai.ML.Visualizers/StateComponentVisualizer.cs:28]: Updated

File: src/Bonsai.ML.Visualizers/TimeSeriesOxyPlotBase.cs (hunk 6 above)
Comment (ncguilbeault) [reply src/Bonsai.ML.Visualizers/TimeSeriesOxyPlotBase.cs:55]: Updated

**Comment by ncguilbeault** [comment]
Thanks a lot for your feedback @glopesdev. I've implemented essentially all of your suggestions and pushed them as new commits.

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (glopesdev) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs]: Ok, finally figured out this is due to tabs versus spaces. It should be possible to run a command in VS code to replace all tabs with spaces for indentations, or just manually find and replace.

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+
+            return $"x0={_x0},x1={_x1},xsteps={_xsteps},y0={_y0},y1={_y1},ysteps={_ysteps}";
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/GridParameters.cs]: Updated

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/KFModelParameters.cs [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+namespace Bonsai.ML.LinearDynamicalSystems.LinearRegression
+{
+
+    /// <summary>
+    /// Model parameters for a Kalman Filter Kinematics python class
+    /// </summary>
+    [Description("Model parameters for a Kalman Filter Linear Regression (KFLR) model")]
+    [Combinator()]
+    [WorkflowElementCategory(ElementCategory.Source)]
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/KFModelParameters.cs]: Updated

File: src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/KFModelParameters.cs [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        private string _xString;
+        private string _pString;
+
+        /// <summary>
+        /// likelihood precision coefficient
+        /// </summary>
+        [JsonProperty("likelihood_precision_coef")]
+        [Description("likelihood precision coefficient")]
+        [Category("Parameters")]
+        public double LikelihoodPrecisionCoef
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.LinearDynamicalSystems/LinearRegression/KFModelParameters.cs]: Updated

File: src/Bonsai.ML.Visualizers/ColorPalettes.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
+namespace Bonsai.ML.Visualizers
+{
+    internal enum ColorPalettes
```
[the remark this answers is not in the document]
Comment (ncguilbeault) [reply src/Bonsai.ML.Visualizers/ColorPalettes.cs]: Updated

**Review by glopesdev — APPROVED** [review]

**Pull request merged**